cities/queries/causal_insight.py

```diff
@@ -0,0 +1,413 @@
+import os
+
+import dill
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import plotly.graph_objects as go
+import pyro
+import torch
+
+from cities.modeling.model_interactions import model_cities_interaction
+from cities.modeling.modeling_utils import (
+    prep_wide_data_for_inference,
+    revert_prediction_df,
+    revert_standardize_and_scale_scaler,
+)
+from cities.utils.cleaning_utils import find_repo_root
+from cities.utils.data_grabber import DataGrabber
+
+
+class CausalInsight:
+    def __init__(self, outcome_dataset, intervention_dataset, num_samples=1000):
+        self.outcome_dataset = outcome_dataset
+        self.intervention_dataset = intervention_dataset
+        self.root = find_repo_root()
+        self.num_samples = num_samples
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    def load_guide(self, forward_shift):
+        pyro.clear_param_store()
+        guide_name = (
+            f"{self.intervention_dataset}_{self.outcome_dataset}_{forward_shift}"
+        )
+        guide_path = os.path.join(
+            self.root, "data/model_guides", f"{guide_name}_guide.pkl"
+        )
+
+        with open(guide_path, "rb") as file:
+            self.guide = dill.load(file)
+        param_path = os.path.join(
+            self.root, "data/model_guides", f"{guide_name}_params.pth"
+        )
+
+        pyro.get_param_store().load(param_path)
+
+        self.forward_shift = forward_shift
+
+    def generate_samples(self):
+        self.data = prep_wide_data_for_inference(
```

Comment:
So a lot of these attributes are defined outside of the init, which can get a little bit funky — "standard" practice is to define all your attributes in the __init__ (setting yet-undefined things like so: `self.value: Optional[Blabla] = None`).

TLDR for rationale is just that it's nice to have all the attributes of a class listed out in one place — is as much for documentation as much as standardizing your error checking to be checking for `None` and not catching attribute errors.

Btw: it's probably worth throwing some errors if one of these methods you expect to have been called hasn't yet been called. That way you can clarify directly that x method should have been called, and future you won't be wondering why you're getting a random attribute error ;)

Reply:
That said, see file level comment.

Reply:
Generally, I get the idea that attributes should be introduced upfront. However, here, the more involved attributes often require some computations and aren't always needed for all uses of an object so producing them upfront would be wasteful.  This is important especially since we care about speed and memory on the front end side.

Reply:
I would say then just set the attributes to None in the `__init__` as a form of documentation.